### PR TITLE
CoverCrypt hybridization

### PR DESCRIPTION
**Pull request by tbrezot**
Hybridize CoverCrypt scheme using CPA-Kyber (version optimized in [the paper](https://github.com/Cosmian/covercrypt)):
```
Encaps(mpk, target_set, K):
    r <- random_scalar();
    C <- U * r;
    D <- V * r;
    E <- HashSet::new();
    for partition in target_set:
        for (kyber_pk_i, H_i) in mpk:
            K_i <- H_i * r;
            // derive the tag for the EAKEM
            K1_i || K2_i <- KDF(K_i, info);
            E_i <- K xor K2_i;
            // implements optimized hybridization:
            // use CPA-Kyber to encrypt `E_i`
            EPQ_i <- CPA-Kyber.Enc(kyber_pk_i, E_i);    // E_i length needs to be 32 bytes
            E.insert((K1_i, EPQ_i));
    return { C, D, E};

Decaps(usk, encapsulation):
    precomp <- encapsulation.C * usk.a + encapsulation.D * usk.b;
    for (K1_i, EPQ_i) in encapulation.E;
        for (kyber_sk_j, x_j) in usk.x:
        K_j <- precomp * x_j;
        // compute the EAKEM tag
        K1_j || K2_j <- KDF(K_j);
        if K1_i == K1_j:
            // the subkey `x_j` is in the target set of the encapsulation
            // `K2_i = K2_j` and `E_j = E_i`
            E_j <- CPA-Kyber.Dec(kyber_sk_j, EPQ_i);
            K <- E_i xor K2_j;
            return K;
    return Err(Authorization)
```

The hybridized version of CoverCrypt is used if the `hybrid` feature is passed. The pre-quantum version will be used otherwise.

**TODO**: `CPA-Kyber.Enc()` takes a `coin` as parameter -> what should be taken ?

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -41,9 +42,11 @@ serde_json = "1.0"
 sha3 = "0.10"
 thiserror = "1.0"
 zeroize = "1.5"
+
 # Optional ones
 js-sys = { version = "0.3", optional = true }
 lazy_static = { version = "1.4", optional = true }
+pqc_kyber = { git = "https://github.com/tbrezot/kyber.git", branch = "feat/expose_indcpa_primitives", features = ["wasm", "ind_cpa"], optional = true}
```
Comment (ThibaudDauce) [inline Cargo.toml:49]: Did you open a PR on the main project about that?

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,9 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 
+- `hybid` feature
```
Comment (Manuthor) [inline CHANGELOG.md:11]:
```suggestion
- `hybrid` feature
```

File: Cargo.toml (hunk 1 above)
Comment (Manuthor) [reply Cargo.toml:49]: we won't be able to publish the crate with a `git` dependency, it does matter for KMS

File: src/interfaces/ffi/tests.rs [hunk 3]
```diff
@@ -66,10 +66,9 @@ unsafe fn encrypt_header(
         authenticated_data.as_ref().len() as i32,
     ))?;
 
-    let symmetric_key_ = SymmetricKey::try_from_bytes(std::slice::from_raw_parts(
-        symmetric_key_ptr.cast(),
-        symmetric_key_len as usize,
-    ))
+    let symmetric_key_ = <DEM as Dem<{ DEM::KEY_LENGTH }>>::Key::try_from_bytes(
```
Comment (Manuthor) [inline src/interfaces/ffi/tests.rs:69]: seems a little bit hard to read

**Review by Manuthor — COMMENTED** [review]
do you keep somewhere the description you gave in this PR? potentially in `cover_crypt_core.rs`?

File: Cargo.toml (hunk 1 above)
Comment (tbrezot) [reply Cargo.toml:48]: I didn't open a PR, but an issue before forking. I guess I could open a PR now.

**Pull request closed without merging**